pymc4/flow/transformed_executor.py

```diff
@@ -0,0 +1,79 @@
+import copy
+
+from pymc4 import scopes, distributions
+from pymc4.distributions import abstract
+from pymc4.distributions.abstract.transforms import JacobianPreference
+from pymc4.flow import SamplingExecutor
+
+
+class TransformedSamplingExecutor(SamplingExecutor):
+    def modify_distribution(self, dist, model_info, state):
```

Comment:
This code here is pretty hard to follow and seems rather complex, sure this is the right design?

Reply:
The source of complexity comes from the following desirable use cases:

1. You want to get the initial point in the transformed space. Here you usually provide test values in the untransformed space and expect transform to be applied. That is what in
https://github.com/pymc-devs/pymc4/pull/129/files/5ec9546526c26ef7eac454f43bcc48951b4d479e#diff-38d8573a4755732b98ab055cf77822adR29
done

```
given: untransformed
do:
    transform
    record transformed
    record logdet
    return untransformed
end
```

2. You actually want to sample and provide values in the transformed space
https://github.com/pymc-devs/pymc4/pull/129/files/5ec9546526c26ef7eac454f43bcc48951b4d479e#diff-38d8573a4755732b98ab055cf77822adR23
Probably we would exclude the case when both values (in the transformed and untransformed case are passed). But it is proceeded silently for now. Maybe warning?
```
given: transformed
do:
    inverse transform
    record untransformed
    record logdet
    return untransformed
end
```

3. You want to sample in the transformed space, the same purpose as in 1., to get initial values for a chain.

```
given: nothing
do:
    sample untransformed 
    # ^--here we are supposed to yield or manually call .sample()
    # I decided to call yield as additional safety checks are performed
    record untransformed
    transform
    record transformed
    record logdet
    return untransformed
end
```

The most tricky part is probably sampling, you can't just sample, you need to record it into the state dict after transform applied. But you first apply the transform and expect the user to get the return value. So sampling comes first, transform second. You see, that differs from 1 and 2 case, when we did no sampling at all.

Reply:
In pymc3 we had a concept of transformed distributions
https://github.com/pymc-devs/pymc3/blob/master/pymc3/model.py#L1557
We manually applied one to the random variable. Do we need the same and how does this look like in here for the same purpose? The very first challenge I got thinking about it is how to decide on the 3 cases described above. Finally you may come up with mixed transformed and untransformed distributions in the same state, but the difference is only `logdet` between them. You should additionally think about conventions "when to keep one or another" and "how to check for duplicates". After a clear picture was not obtained, I decided on a dumb idea. If you want to compute logp for the transformed space, you'll see logdets added, the rest computations and rules are the same.

Reply:
Thanks for the detailed explanation, unfortunately I'll be traveling the rest of the week and not sure I can read this in detail soon, but do remind me next week if I don't get back to you. Or maybe @ColCarroll and @junpenglao have thoughts.